celeri/celeri.py

```diff
@@ -498,6 +500,157 @@ def process_segment(segment, command, meshes):
+        # Final selection
+        top_edge_indices = np.sort(np.hstack((top_edge_indices1, top_edge_indices2)))
+        # Get new segment coordinates from these indices
+        edge_segs = make_default_segment(len(top_edge_indices))
+        edge_segs.lon1 = meshes[i].meshio_object.points[
+            meshes[i].ordered_edge_nodes[top_edge_indices, 0], 0
+        ]
+        edge_segs.lat1 = meshes[i].meshio_object.points[
+            meshes[i].ordered_edge_nodes[top_edge_indices, 0], 1
+        ]
+        edge_segs.lon2 = meshes[i].meshio_object.points[
+            meshes[i].ordered_edge_nodes[top_edge_indices, 1], 0
+        ]
+        edge_segs.lat2 = meshes[i].meshio_object.points[
+            meshes[i].ordered_edge_nodes[top_edge_indices, 1], 1
+        ]
+        edge_segs.locking_depth = -15
+        edge_segs.patch_flag = +1
+        edge_segs.patch_file_name = +i + 1
+        all_edge_segment = all_edge_segment.append(edge_segs)
+
+    # Get indices of segments to keep
+    keep_segment_idx = np.setdiff1d(range(len(segment.lon1)), cut_segment_idx)
+    # Isolate kept segments and reindex
+    keep_segment = segment.loc[keep_segment_idx]
+    new_index = range(len(keep_segment_idx))
+    keep_segment.index = new_index
+    # Find hanging endpoints; these mark terminations of mesh-replaced segments
+    lons = np.hstack((keep_segment.lon1, keep_segment.lon2))
+    lats = np.hstack((keep_segment.lat1, keep_segment.lat2))
+    coords = np.array([lons, lats])
+    unique_coords, indices, counts = np.unique(
+        coords, axis=1, return_index=True, return_counts=True
+    )
+    hanging_idx = indices[np.where(counts == 1)]
+    # Calculate distance to all mesh edge coordinates
+    # Can't just use the terminations because we might have triple junctions in the middle of a mesh
+    elons = np.hstack((all_edge_segment.lon1, all_edge_segment.lon2))
+    elats = np.hstack((all_edge_segment.lat1, all_edge_segment.lat2))
+    ecoords = np.array([elons, elats])
+    hang_to_mesh_dist = cdist(coords[:, hanging_idx].T, ecoords.T)
+    # Find closest edge coordinate
+    closest_edge_idx = np.argmin(hang_to_mesh_dist, axis=1)
+    # Replace segment coordinates with closest mesh coordinate
+    # Using a loop because we need to evaluate whether to replace endpoint 1 or 2
+    for i in range(len(closest_edge_idx)):
+        if hanging_idx[i] < len(keep_segment.lon1):
+            keep_segment.loc[hanging_idx[i], "lon1"] = ecoords[0, closest_edge_idx[i]]
+            keep_segment.loc[hanging_idx[i], "lat1"] = ecoords[1, closest_edge_idx[i]]
+        else:
+            keep_segment.loc[hanging_idx[i] - len(keep_segment.lon1), "lon2"] = ecoords[
+                0, closest_edge_idx[i]
+            ]
+            keep_segment.loc[hanging_idx[i] - len(keep_segment.lon1), "lat2"] = ecoords[
+                1, closest_edge_idx[i]
+            ]
+    # Merge with mesh edge segments
+    new_segment = keep_segment.append(all_edge_segment)
```

Comment:
@jploveless Reading through this now.  Should appending happen for the Cascadia case?  If so do we need to delete the list elements with the previous coordinates?